File: src/D01.cs
namespace src01
{
  using System;
  using System.Collections.Generic;
  using System.Linq;
  using System.Reactive.Linq;

  public static class Code
  {
    public static int FuelForMass(int mass) => mass/3 - 2;
    public static int TotalFuelForMass(int mass) 
     => Observable.Generate(mass, m => m > 0, m => FuelForMass(m), m => m).ToEnumerable().Skip(1).Sum();

    public static int SumOf(this IEnumerable<string> input, Func<int,int> getValue)
      => input.Select(l => getValue(int.Parse(l))).Sum();

  }
}

File: srcf/D01.fs
module Src01

let FuelForMass mass =
  mass/3 - 2

let TotalFuelForMass mass =
  Seq.unfold (fun m -> Some(m,FuelForMass m)) mass
    |> Seq.takeWhile (fun x -> x > 0)
    |> Seq.skip 1
    |> Seq.sum

let SumOf  (f : int -> int) (l : string seq) =
  l |> Seq.sumBy(int >> f)


